Update asset importing spec to use `import.meta.resolve` (#973)
emberjs/rfcs: text/0763-asset-importing-spec.md

--------------------|-----------------------|--------|---
JS | JS | `import { stuff } from 'thing';`| ECMA with NodeJS resolving rules
JS | CSS | `import 'thing/style.css';` | [Embroider Spec CSS Rule](https://github.com/emberjs/rfcs/blob/master/text/0507-embroider-v2-package-format.md#css) (1)
JS| Other Asset (2) | `import.meta.resolve('thing/icon.png');`| **Proposed** URL Rule (3)
CSS | CSS | `@import("thing/style.css")` | [W3C](https://drafts.csswg.org/css-cascade/#at-import) plus NodeJS resolving rules
CSS | Other | `url("thing/icon.png");` | W3C plus NodeJS resolving rules
HTML | JS | `<script src="./thing.js"></script>` | W3C
HTML | CSS | `<link rel="stylesheet" href="./thing.css"/>` | W3C
HTML | Other | `<img src="./thing.png" />` | W3C

1. Importing a file with an explicit `.css` extension guarantees that the given CSS will be loaded into the DOM before your module executes. We do not define any exported values, this is purely for side-effect. This rule is not part of the present RFC, it was already in [RFC 507](https://github.com/emberjs/rfcs/blob/master/text/0507-embroider-v2-package-format.md).

> 2. "Other Asset" means this rule applies to any JS import of a path that ends in an explicit file extension that is not `.js` or `.css`, meaning:

Author: ```suggestion
2. "Other Asset" means this rule applies to any `import.meta.resolve` with a path that ends in an explicit file extension that is not `.js` or `.css`, meaning:
```

### `import.meta.resolve(moduleName)`

`import.meta.resolve(moduleName)`[^1] is a function built-in to browsers that takes a `moduleName` and "returns a string corresponding to the path that would be imported if the argument were passed to import()." [^2]

The build system will handle uses of `import.meta.resolve` when called with assets other than `.js` or .css` (see above).

The argument `moduleName` supports a subset of dynamicism as defined [here](https://github.com/emberjs/rfcs/blob/master/text/0507-embroider-v2-package-format.md#supported-subset-of-dynamic-import-syntax).

The argument `moduleName` must explicitly include the extension.

For example, the following is valid as the build system can determine the files that possibly match the interpolation:

```js
class extends Component {
  get iconURL() {
    return import.meta.resolve(`../icons/${this.args.locale}.png`)
  }
}
```

Calling `import.meta.resolve` with an asset that does not exist will cause a build error. You cannot fail to notice if you accidentally delete an asset that is still in use somewhere in your app.

#### Isn't it risky to "take over" a browser API?

> Yes and no. This [original proposal for this feature to WHATWG](https://github.com/whatwg/html/issues/3871#issue-346547968) envisioned build tools rewriting the expressions to provide asset optimizations.

Reviewer: What's the risk that "Yes" is referring to? I read the following text as arguing against the existence of risk, or do I misunderstand this?

Author: No, just poor wording.

> Yes and no. This [original proposal for this feature to WHATWG](https://github.com/whatwg/html/issues/3871#issue-346547968) envisioned build tools rewriting the expressions to provide asset optimizations.

Author: ```suggestion
No. This [original proposal for this feature to WHATWG](https://github.com/whatwg/html/issues/3871#issue-346547968) envisioned build tools rewriting the expressions to provide asset optimizations. As describe above, it is only taken over for "Other Assets" (any path that ends in an explicit file extension that is not `.js` or `.css`). 
```
